File: jdbc/src/test/java/net/snowflake/jdbc/e2e/put_get/PutGetBasicOperationsTest.java
package net.snowflake.jdbc.e2e.put_get;

import static net.snowflake.jdbc.utils.DriverCompatibility.isNewDriver;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.nio.file.Files;
import java.nio.file.Path;
import java.sql.Connection;
import java.sql.ResultSet;
import java.sql.ResultSetMetaData;
import java.sql.Statement;
import java.sql.Types;
import java.util.List;
import net.snowflake.jdbc.utils.SnowflakeIntegrationTestBase;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.io.TempDir;

public class PutGetBasicOperationsTest extends SnowflakeIntegrationTestBase implements WithPutGet {

  private Path testDataCsv() {
    return sharedTestDataDir().resolve("compression").resolve("test_data.csv");
  }

  @Test
  public void shouldSelectDataFromFileUploadedToStage() throws Exception {
    // Given File is uploaded to stage
    String stageName =
        createStageAndUploadFile(
            getDefaultConnection(), "TEST_STAGE_SELECT", testDataCsv(), true, true);

    // When File data is queried using Select command
    try (Statement statement = getDefaultConnection().createStatement();
        ResultSet resultSet = statement.executeQuery("SELECT $1, $2, $3 FROM @" + stageName)) {
      // Then File data should be correctly returned
      assertTrue(resultSet.next(), "Expected one row from the uploaded file");
      assertEquals("1", resultSet.getString(1), "Unexpected first column");
      assertEquals("2", resultSet.getString(2), "Unexpected second column");
      assertEquals("3", resultSet.getString(3), "Unexpected third column");
      assertFalse(resultSet.next(), "Expected exactly one row");
    }
  }

  @Test
  public void shouldListFileUploadedToStage() throws Exception {
    // Given File is uploaded to stage
    String stageName =
        createStageAndUploadFile(
            getDefaultConnection(), "TEST_STAGE_LS", testDataCsv(), true, true);

    // When Stage content is listed using LS command
    List<String> files = listStageFileNames(getDefaultConnection(), stageName);

    // Then File should be listed with correct filename
    assertEquals(1, files.size(), "Expected exactly one file on the stage");
    assertTrue(
        files.get(0).contains("test_data.csv.gz"),
        "Expected the compressed filename, got: " + files.get(0));
  }

  @Test
  public void shouldGetFileUploadedToStage(@TempDir Path downloadDir) throws Exception {
    // Given File is uploaded to stage
    String stageName =
        createStageAndUploadFile(
            getDefaultConnection(), "TEST_STAGE_GET", testDataCsv(), true, true);

    // When File is downloaded using GET command
    List<GetRow> rows =
        getFileFromStage(getDefaultConnection(), stageName, "test_data.csv", downloadDir);

    // Then File should be downloaded
    assertEquals(1, rows.size(), "Expected exactly one downloaded file");
    assertEquals("DOWNLOADED", rows.get(0).status, "Expected DOWNLOADED status");
    Path downloaded = downloadDir.resolve("test_data.csv.gz");
    assertTrue(Files.exists(downloaded), "Expected the compressed file on disk");

    // And Have correct content
    assertEquals("1,2,3", readTextMaybeGzip(downloaded).trim(), "Unexpected downloaded content");
  }

  @Test
  public void shouldReturnCorrectRowsetForPut() throws Exception {
    // Given Snowflake client is logged in
    Connection connection = getDefaultConnection();

    // When File is uploaded to stage
    String stageName = createTemporaryStage(connection, "TEST_STAGE_PUT_ROWSET");
    PutRow row = uploadFileToStage(connection, stageName, testDataCsv(), true, true);

    // Then Rowset for PUT command should be correct
    assertEquals("test_data.csv", row.source, "Unexpected source filename");
    assertEquals("test_data.csv.gz", row.target, "Unexpected target filename");
    assertEquals(6, row.sourceSize, "Unexpected source size");
    assertEquals(26, row.targetSize, "Unexpected target size");
    assertEquals("NONE", row.sourceCompression, "Unexpected source compression");
    assertEquals("GZIP", row.targetCompression, "Unexpected target compression");
    assertEquals("UPLOADED", row.status, "Unexpected status");
    assertEquals("ENCRYPTED", row.encryption, "Unexpected encryption");
    assertEquals("", row.message, "Expected empty message");
  }

  @Test
  public void shouldReturnCorrectRowsetForGet(@TempDir Path downloadDir) throws Exception {
    // Given File is uploaded to stage
    String stageName =
        createStageAndUploadFile(
            getDefaultConnection(), "TEST_STAGE_GET_ROWSET", testDataCsv(), true, true);

    // When File is downloaded using GET command
    List<GetRow> rows =
        getFileFromStage(getDefaultConnection(), stageName, "test_data.csv", downloadDir);

    // Then Rowset for GET command should be correct
    assertEquals(1, rows.size(), "Expected exactly one downloaded file");
    GetRow row = rows.get(0);
    assertEquals("test_data.csv.gz", row.file, "Unexpected downloaded filename");
    assertEquals(26, row.size, "Unexpected downloaded size");
    assertEquals("DOWNLOADED", row.status, "Unexpected status");
    assertEquals("DECRYPTED", row.encryption, "Unexpected encryption");
    assertEquals("", row.message, "Expected empty message");
  }

  @Test
  public void shouldReturnCorrectColumnMetadataForPut() throws Exception {
    // Given Snowflake client is logged in
    Connection connection = getDefaultConnection();

    // When File is uploaded to stage
    String stageName = createTemporaryStage(connection, "TEST_STAGE_PUT_COLUMN_METADATA");
    String putSql =
        String.format(
            "PUT '%s' @%s AUTO_COMPRESS=TRUE OVERWRITE=TRUE", fileUri(testDataCsv()), stageName);
    try (Statement statement = connection.createStatement();
        ResultSet resultSet = statement.executeQuery(putSql)) {
      // Then Column metadata for PUT command should be correct
      ResultSetMetaData meta = resultSet.getMetaData();
      assertEquals(9, meta.getColumnCount(), "PUT should return 9 columns");
      assertTextColumn(meta, 1, "source");
      assertTextColumn(meta, 2, "target");
      assertFixedColumn(meta, 3, "source_size");
      assertFixedColumn(meta, 4, "target_size");
      assertTextColumn(meta, 5, "source_compression");
      assertTextColumn(meta, 6, "target_compression");
      assertTextColumn(meta, 7, "status");
      assertTextColumn(meta, 8, "encryption");
      assertTextColumn(meta, 9, "message");
      assertTrue(resultSet.next(), "Expected one PUT result row");
      assertEquals("UPLOADED", resultSet.getString(7), "Expected UPLOADED status");
      assertFalse(resultSet.wasNull(), "status should be non-null");
    }
  }

  @Test
  public void shouldReturnCorrectColumnMetadataForGet(@TempDir Path downloadDir) throws Exception {
    // Given File is uploaded to stage
    String stageName =
        createStageAndUploadFile(
            getDefaultConnection(), "TEST_STAGE_GET_COLUMN_METADATA", testDataCsv(), true, true);

    // When File is downloaded using GET command
    String getSql = String.format("GET @%s/test_data.csv '%s/'", stageName, fileUri(downloadDir));
    try (Statement statement = getDefaultConnection().createStatement();
        ResultSet resultSet = statement.executeQuery(getSql)) {
      // Then Column metadata for GET command should be correct
      ResultSetMetaData meta = resultSet.getMetaData();
      assertEquals(5, meta.getColumnCount(), "GET should return 5 columns");
      assertTextColumn(meta, 1, "file");
      assertFixedColumn(meta, 2, "size");
      assertTextColumn(meta, 3, "status");
      assertTextColumn(meta, 4, "encryption");
      assertTextColumn(meta, 5, "message");
      assertTrue(resultSet.next(), "Expected one GET result row");
      assertEquals("DOWNLOADED", resultSet.getString(3), "Expected DOWNLOADED status");
      assertFalse(resultSet.wasNull(), "status should be non-null");
    }
  }

  private static void assertTextColumn(ResultSetMetaData meta, int column, String expectedName)
      throws Exception {
    assertEquals(
        expectedName,
        meta.getColumnName(column).toLowerCase(),
        "Unexpected name for column " + column);
    assertEquals(
        Types.VARCHAR, meta.getColumnType(column), "Expected TEXT column at position " + column);
  }

  private static void assertFixedColumn(ResultSetMetaData meta, int column, String expectedName)
      throws Exception {
    assertEquals(
        expectedName,
        meta.getColumnName(column).toLowerCase(),
        "Unexpected name for column " + column);
    // BD#54: the new driver reports PUT/GET size columns as BIGINT (standard scale-0 rule);
    // legacy hard-codes them to DECIMAL.
    int expectedType = isNewDriver() ? Types.BIGINT : Types.DECIMAL;
    assertEquals(
        expectedType, meta.getColumnType(column), "Expected FIXED column at position " + column);
  }
}

File: nodejs/tests/e2e/query/put-get/put-get-basic-operations.test.ts
import fs from 'node:fs';
import path from 'node:path';
import zlib from 'node:zlib';
import { describe, it, expect, beforeAll, afterAll, beforeEach, afterEach } from 'vitest';
import type { Connection, RowStatement } from '../../../types/sdk-types.js';
import { sharedTestDataDir, toFileUrl } from '../../utils/files.js';
import { createLiveConnection, createTempDir } from '../../utils/fixtures.js';
import {
  destroyConnectionAsync,
  executeAsync,
  expectColumnsNames,
  isRunningNewDriverWithBD,
  randomizeName,
} from '../../utils/index.js';

const PUT_COLUMNS = [
  'source',
  'target',
  'sourceSize',
  'targetSize',
  'sourceCompression',
  'targetCompression',
  'status',
  'message',
];
const GET_COLUMNS = ['file', 'size', 'status', 'message'];

// TODO: duplicates put-get.test.ts but doesn't test all compression cases
// maybe they should be merged
describe('PUT GET basic operations', () => {
  let connection: Connection;

  beforeAll(async () => {
    connection = await createLiveConnection({}, false);
  });

  afterAll(async () => {
    await destroyConnectionAsync(connection);
  });

  describe('tests/definitions/shared/put_get/put_get_basic_operations.feature', () => {
    const TEST_FILE = path.join(sharedTestDataDir(), 'compression', 'test_data.csv');

    let stage: string;

    beforeEach(async () => {
      stage = randomizeName('TEST_STAGE_BASIC');
      await executeAsync(connection, `CREATE TEMPORARY STAGE IF NOT EXISTS ${stage}`);
    });

    afterEach(async () => {
      await executeAsync(connection, `DROP STAGE IF EXISTS ${stage}`);
    });

    it('should select data from file uploaded to stage', async () => {
      // Given File is uploaded to stage
      await executeAsync(
        connection,
        `PUT ${toFileUrl(TEST_FILE)} @${stage} AUTO_COMPRESS=TRUE OVERWRITE=TRUE`,
      );

      // When File data is queried using Select command
      const { rows } = await executeAsync(connection, `SELECT $1, $2, $3 FROM @${stage}`);

      // Then File data should be correctly returned
      expect(Object.values(rows[0])).toEqual(['1', '2', '3']);
    });

    it('should list file uploaded to stage', async () => {
      // Given File is uploaded to stage
      await executeAsync(
        connection,
        `PUT ${toFileUrl(TEST_FILE)} @${stage} AUTO_COMPRESS=TRUE OVERWRITE=TRUE`,
      );

      // When Stage content is listed using LS command
      const { rows } = await executeAsync(connection, `LS @${stage}`);

      // Then File should be listed with correct filename
      expect(rows).toHaveLength(1);
      expect(rows[0].name).toContain('test_data.csv.gz');
    });

    it('should get file uploaded to stage', async () => {
      // Given File is uploaded to stage
      await executeAsync(
        connection,
        `PUT ${toFileUrl(TEST_FILE)} @${stage} AUTO_COMPRESS=TRUE OVERWRITE=TRUE`,
      );

      const downloadDir = createTempDir();
      // When File is downloaded using GET command
      const { rows } = await executeAsync(
        connection,
        `GET @${stage}/test_data.csv ${downloadDir.fileUrl()}/`,
      );

      // Then File should be downloaded
      expect(rows[0].status).toBe('DOWNLOADED');
      const downloadedFile = downloadDir.resolve('test_data.csv.gz');
      expect(fs.existsSync(downloadedFile)).toBe(true);

      // And Have correct content
      expect(zlib.gunzipSync(fs.readFileSync(downloadedFile)).toString().trim()).toBe('1,2,3');
    });

    it('should return correct rowset for PUT', async () => {
      // Given Snowflake client is logged in
      void connection;

      // When File is uploaded to stage
      const { rows } = await executeAsync(
        connection,
        `PUT ${toFileUrl(TEST_FILE)} @${stage} AUTO_COMPRESS=TRUE OVERWRITE=TRUE`,
      );

      // Then Rowset for PUT command should be correct
      expect(rows[0]).toMatchObject({
        source: 'test_data.csv',
        target: 'test_data.csv.gz',
        sourceSize: 6,
        targetSize: isRunningNewDriverWithBD('BD#23') ? 32 : 26,
        sourceCompression: isRunningNewDriverWithBD('BD#24') ? 'NONE' : null,
        targetCompression: 'GZIP',
        status: 'UPLOADED',
        message: isRunningNewDriverWithBD('BD#25') ? '' : undefined,
      });
    });

    it('should return correct rowset for GET', async () => {
      // Given File is uploaded to stage
      await executeAsync(
        connection,
        `PUT ${toFileUrl(TEST_FILE)} @${stage} AUTO_COMPRESS=TRUE OVERWRITE=TRUE`,
      );

      const downloadDir = createTempDir();
      // When File is downloaded using GET command
      const { rows } = await executeAsync(
        connection,
        `GET @${stage}/test_data.csv ${downloadDir.fileUrl()}/`,
      );

      // Then Rowset for GET command should be correct
      expect(rows[0]).toMatchObject({
        file: 'test_data.csv.gz',
        size: 26,
        status: 'DOWNLOADED',
        message: isRunningNewDriverWithBD('BD#25') ? '' : undefined,
      });
    });

    it('should return correct column metadata for PUT', async () => {
      // Given Snowflake client is logged in
      void connection;

      // When File is uploaded to stage
      const { statement, rows } = await executeAsync(
        connection,
        `PUT ${toFileUrl(TEST_FILE)} @${stage} AUTO_COMPRESS=TRUE OVERWRITE=TRUE`,
      );

      // Then Column metadata for PUT command should be correct
      expectColumnsNames(statement as RowStatement, PUT_COLUMNS);
      expect(rows[0].status).toBe('UPLOADED');
    });

    it('should return correct column metadata for GET', async () => {
      // Given File is uploaded to stage
      await executeAsync(
        connection,
        `PUT ${toFileUrl(TEST_FILE)} @${stage} AUTO_COMPRESS=TRUE OVERWRITE=TRUE`,
      );

      const downloadDir = createTempDir();
      // When File is downloaded using GET command
      const { statement, rows } = await executeAsync(
        connection,
        `GET @${stage}/test_data.csv ${downloadDir.fileUrl()}/`,
      );

      // Then Column metadata for GET command should be correct
      expectColumnsNames(statement as RowStatement, GET_COLUMNS);
      expect(rows[0].status).toBe('DOWNLOADED');
    });

    it('should upload file to subdirectory in stage', async () => {
      // Given Snowflake client is logged in
      void connection;

      // When File is uploaded to a subdirectory in stage
      const { rows: uploaded } = await executeAsync(
        connection,
        `PUT ${toFileUrl(TEST_FILE)} @${stage}/nested/subdir AUTO_COMPRESS=FALSE`,
      );
      expect(uploaded[0].status).toBe('UPLOADED');

      // Then File should be listed under the subdirectory
      const { rows } = await executeAsync(connection, `LS @${stage}`);
      expect(
        rows.some(
          (row) =>
            String(row.name).includes('nested/subdir') &&
            String(row.name).includes('test_data.csv'),
        ),
      ).toBe(true);
    });

    it('should get file from subdirectory in stage', async () => {
      // Given File is uploaded to a subdirectory in stage
      const { rows: uploaded } = await executeAsync(
        connection,
        `PUT ${toFileUrl(TEST_FILE)} @${stage}/nested/subdir AUTO_COMPRESS=FALSE OVERWRITE=TRUE`,
      );
      expect(uploaded[0].status).toBe('UPLOADED');

      const downloadDir = createTempDir();
      // When All files are downloaded from stage using GET command
      const { rows } = await executeAsync(connection, `GET @${stage}/ ${downloadDir.fileUrl()}/`);

      // Then File should be downloaded flat into the local directory
      expect(rows[0].status).toBe('DOWNLOADED');
      const downloadedFile = downloadDir.resolve('test_data.csv');
      expect(fs.existsSync(downloadedFile)).toBe(true);

      // And Have correct content
      expect(fs.readFileSync(downloadedFile, 'utf8').trim()).toBe('1,2,3');
    });
  });
});
